packages/core/usePrevious/index.ts

```diff
@@ -0,0 +1,32 @@
+import { type ComputedRef, type Ref, ref, watchEffect } from 'vue-demi'
+
+/**
+ * Updates prevRef when true is returned
+ * default (a?: T, b?: T) => !Object.is(a, b)
+ */
+export type ShouldUpdateFunc<T> = (prev: T | undefined, next: T) => boolean
+
+const defaultShouldUpdate = <T>(a?: T, b?: T) => !Object.is(a, b)
+
+/**
+ * A Hook to return the previous state.
+ *
+ * @see https://vueuse.org/core/usePrevious
+ * @param state
+ * @param [shouldUpdate]
+ */
+export function usePrevious<T>(
+  state: Ref<T> | ComputedRef<T>,
+  shouldUpdate: ShouldUpdateFunc<T> = defaultShouldUpdate,
+) {
+  const prevRef = ref<T>()
+  const curRef = ref<T>()
+  watchEffect(() => {
+    if (shouldUpdate(curRef.value, state.value)) {
+      prevRef.value = curRef.value
+      curRef.value = state.value
+    }
+  })
```

Comment:
Seems `curRef` is just a intermediate state that doesn't necessarily need to be reactive? would slightly improve the performance.
```suggestion
  let cur: T | undefined
  watchEffect(() => {
    if (shouldUpdate(cur, state.value)) {
      prevRef.value = cur
      cur = state.value
    }
  })
```

Reply:
sorry being verbose